### PR TITLE
Init objectset apis

### PR DESCRIPTION
**Pull request by thetechnick**

### CODE REVIEW

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	ObjectSetTemplateSpec `json:",inline"`
+}
+
+// ClusterObjectSetStatus defines the observed state of a ClusterObjectSet.
+type ClusterObjectSetStatus struct {
+	// Conditions is a list of status conditions ths object is in.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// Deprecated: This field is not part of any API contract
+	// it will go away as soon as kubectl can print conditions!
+	// Human readable status - please use .Conditions from code
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:55]:
```suggestion
	// For human readable status - please use .Conditions from code
```

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	Phase ObjectSetStatusPhase `json:"phase,omitempty"`
+	// List of objects, the controller has paused reconciliation on.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:57]:
```suggestion
	// List of objects the controller has paused reconciliation on.
```

File: apis/core/v1alpha1/commonobjectset_types.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+const (
+	// "Active" is the default lifecycle state.
+	ObjectSetLifecycleStateActive ObjectSetLifecycleState = "Active"
+	// "Paused" disables reconciliation of the ObjectSet.
+	// Only Status updates will still propagated, but object changes will not be reconciled.
+	ObjectSetLifecycleStatePaused ObjectSetLifecycleState = "Paused"
+	// "Archived" disables reconciliation while also "scaling to zero",
+	// which deletes all objects that are not excluded via the pausedFor property and
+	// removes itself from the owner list of all other objects previously under management.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/commonobjectset_types.go:18]: Are the objects also deleted if there is no other owner? If so I might include that here.

File: apis/core/v1alpha1/commonobjectset_types.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	// Object Name.
+	// +example=example-deployment
+	Name string `json:"name"`
+}
+
+// ObjectSet Condition Types.
+const (
+	// Available indicates that all objects pass their availability probe.
+	ObjectSetAvailable = "Available"
+	// Paused indicates that object changes are no reconciled, but status is still reported.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/commonobjectset_types.go:67]:
```suggestion
	// Paused indicates that object changes are not reconciled, but status is still reported.
```

File: apis/core/v1alpha1/commonobjectset_types.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	Kind string `json:"kind"`
+}
+
+// Defines probe parameters to check parts of a package.
+type Probe struct {
+	Condition   *ProbeConditionSpec   `json:"condition,omitempty"`
+	FieldsEqual *ProbeFieldsEqualSpec `json:"fieldsEqual,omitempty"`
+}
+
+// Checks wether the object reports a condition with given type and status.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/commonobjectset_types.go:134]:
```suggestion
// Checks whether or not the object reports a condition with given type and status.
```

File: apis/core/v1alpha1/groupversion_info.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// The package v1alpha1 contains API Schema definitions for the v1alpha1 version of the core Page Operator API group,
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/groupversion_info.go:1]: This should be "Package Operator API group" right?

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// ClusterObjectSet reconciles a collection of objects across ordered phases and aggregates their status.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:5]:
```suggestion
// ClusterObjectSet reconciles a collection of objects through ordered phases and aggregates their status.
```

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+//
+// ClusterObjectSets behave similar to Kubernetes ReplicaSets, by managing a collection of objects and being itself mostly immutable.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:7]:
```suggestion
// ClusterObjectSets behave similarly to Kubernetes ReplicaSets, by managing a collection of objects and being itself mostly immutable.
```

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+// +kubebuilder:object:root=true
+type ClusterObjectSetList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterObjectSet `json:"items"`
+}
+
+// ClusterObjectSetSpec defines the desired state of a ClusterObjectSet.
+type ClusterObjectSetSpec struct {
+	// Specifies the lifecycle state of the ObjectSet.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:37]: "... of the _Cluster_ObjectSet."?

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+// ClusterObjectSetStatus defines the observed state of a ClusterObjectSet.
+type ClusterObjectSetStatus struct {
+	// Conditions is a list of status conditions ths object is in.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// Deprecated: This field is not part of any API contract
+	// it will go away as soon as kubectl can print conditions!
+	// Human readable status - please use .Conditions from code
+	Phase ObjectSetStatusPhase `json:"phase,omitempty"`
+	// List of objects, the controller has paused reconciliation on.
+	PausedFor []ObjectSetPausedObject `json:"pausedFor,omitempty"`
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:58]: Probably a dumb question, but why do we need `PausedFor` in both the Spec and Status Objects?

File: apis/core/v1alpha1/commonobjectset_types.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	Phases []ObjectSetTemplatePhase `json:"phases"`
+	// Availability Probes check objects that are part of the package.
+	// All probes need to succeed for a package to be considered Available.
+	// Failing probes will prevent the reconciliation of objects in later phases.
+	AvailabilityProbes []ObjectSetProbe `json:"availabilityProbes"`
+}
+
+// ObjectSet reconcile phase.
+type ObjectSetTemplatePhase struct {
+	// Name of the reconcile phase, must be unique within a ObjectSet.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/commonobjectset_types.go:36]:
```suggestion
	// Name of the reconcile phase, the string mapping of a ObjectSetLifecycleState constant. Must be unique within a ObjectSet.
```

File: apis/core/v1alpha1/commonobjectset_types.go (hunk 3 above)
Comment (thetechnick) [reply apis/core/v1alpha1/commonobjectset_types.go:18]: Nope. GC only kicks in, if all owners referenced by the child are deleted.
If an ObjectSet is Archived, another should have taken over, so objects are not deleted.

File: apis/core/v1alpha1/groupversion_info.go (hunk 6 above)
Comment (thetechnick) [reply apis/core/v1alpha1/groupversion_info.go:1]: Yes! :man_facepalming:

File: apis/core/v1alpha1/clusterobjectset_types.go (hunk 10 above)
Comment (thetechnick) [reply apis/core/v1alpha1/clusterobjectset_types.go:58]: Not a dumb question.
It's so the higher-order controller/user can wait till the controller has seen and confirmed that these objects are now indeed paused.
If we don't synchronize, we will have very evil race conditions :)

**Pull request merged**